### ISSUE TITLE
Array Field Validation Not Working Properly

### ISSUE THREAD

**Issue by TheDunco** [payloadcms/payload#2281]
# Bug Report

Array fields don't seem to be validating properly! Even when they return a string, the admin UI will allow the document to be saved without displaying a warning message.

I was able to confirm that other field types like a simple text field fail validation when they are expected to, but not Array fields. I've confirmed that it gets to the `return "..."` step by `console.warrn()`ing right before the return statement, and I can see the console output but it doesn't validate correctly.

## Steps to Reproduce

1. Have an array field with validation
2. Have invalid data in the field (i.e. something that should fail validation)
3. Hit "Save" to validate, and it passes validation when it shouldn't

I was able to verify that a simple field like this does validate as expected...
```typescript
{
    type: 'text',
    name: 'testValidation',
    validate: (value) => {
        if (value === 'test') {
            return 'You cannot use the word "test" in this field';
        }
        return true;
    },
},
```

However, as soon as I turn this into an array field and validate on that array field, validation will get to the return step but it doesn't fail validation when it should.

```typescript
{
    type: 'array',
    name: 'testArray',
    validate: (value) => {
        if (Array.isArray(value) && value?.length > 0) {
            if (value.find((val) => val.testValidation === 'test')) {
                console.warn('Test validation should fail');
                return 'You cannot use "test" in this field';
            }
        }
        return true;
    },
    fields: [
        {
            type: 'text',
            name: 'testValidation',
        },
    ],
},
```
When I hit "Save" with the code above with the word "test" in the array field, I get the `console.warn`ing but _am able to save it_ when I shouldn't be able to -- it should fail validation because the next line returns a string. The confirmation toast pops up green and there are no field validation errors when there clearly should be.

<img width="1706" alt="image" src="https://user-images.githubusercontent.com/44826828/224125308-c7fedb5d-90dd-4aa6-9fb9-a6e3c4793627.png">

## Other Details

Payload Version: <1.16.11 -> ^1.16.19

**Comment by denolfe** [comment]
The `value` that is being passed is actually _the length of the array_. If you're looking for the actual array values, you can pass a second argument and interrogate the `siblingData` property with the array's slug.

Here is how to re-write your validation function in this way:

```ts
validate: (value, { siblingData }) => {
  const arrayData = siblingData?.testArray;
  if (value && arrayData) {
    if (arrayData.find((val) => val.testValidation === 'test')) {
      console.warn('Test validation should fail');
      return 'You cannot use "test" in this field';
    }
  }
  return true;
},
```

**Comment by TheDunco** [comment]
Roger that, thanks! I was able to get this to work properly. 

@denolfe Is it documented anywhere that the `value` of an array field will be the length of the array? I just want to know if I missed it, an/or where it's documented.

**Comment by denolfe** [comment]
I don't believe so. The docs for validate are very general. This is unique to arrays, and unfortunately is not typed as strongly as it should be at this point in time.

**Comment by TheDunco** [comment]
> I don't believe so. The docs for validate are very general. This is unique to arrays, and unfortunately is not typed as strongly as it should be at this point in time.

Ok, thanks for the info! Is it on the roadmap to eventually type these more strongly?

**Comment by TheDunco** [comment]
Also, I did notice that occasionally `value` would represent the actual value of the array. When actually validating it is the length of the array, but I did see it occasionally come through when logging it as the actual current value of the array which confused me greatly.

This also doesn't explain why I was able to see that I was able to see the `console.log` right before the return statement which was the other source of my confusion. I was able to get it working, though, so this is just for future consideration.
